### DISCUSSION TITLE
Contribution Guidelines

### DISCUSSION THREAD

**Discussion by Bond-Addict** [codemerx/CodemerxDecompile#120] [Q&A]
Hey there! I'm very interested in being a contributor.  I'm unable to locate your contribution guidelines/ process. If I missed it, could you link me to it? If not, could you provide them to me? JustDecompile and JustAssembly used to be my favorite and most commonly used toolset. I'd love to help with the resurrection and with community suggestions when possible 🙂

A bit of a background, I'm a 27 year old Full -Stack .Net Developer that been developing professionally for just under 8 years. I've worked in both the public and private sectors on legacy projects than include bit are limited to Healthcare products, internal and eternal state government project, iot projects and currently work for a health insurance provider. I've worked with Coldfusion, c#, vb.net, vb 6, .Net framework all the way up to .Net 8. I've worked with Frontends like angular and vue.js but also Blazor and am beginning to dabble with Maui for building native apps. I also am quite familiar with both sql server and MongoDb.

This won't be my first open source project to contribute to either. I started using Vuestic-Ui and have made many contributions to their project that I've been able to get a ton of benefit from which include missing features that are now widely used by other developers. 

My #1 priority is always to write clean, self documenting code that to the best of my ability follow the lasted industry standard best practices. 

I haven't touched every available framework or language, but I'm quick study and would love to opportunity to take on the challenge of another project that I have a deep appreciation for. 

Thanks 😃

**Comment by tsviatko** [comment]
Hey, sorry about the late reply.

There are not many contributors as of now so we are pretty flexible. There does not seem to be any need for formal process at this point. 

So if you are willing to help I guess the first thing we should do is pick an issue for you to work on. Then we can answer your questions and give feedback on your work on the fly. 

What would be most interesting for you to work on? Maybe the last one you created - https://github.com/codemerx/CodemerxDecompile/discussions/119 ? Or maybe you would like more something decompilation related? https://github.com/codemerx/CodemerxDecompile/issues/110 is pretty well defined at this point.

**Reply by Bond-Addict** [reply]
#110 sounds pretty interesting. #119 was actually created by me 🤣 Is there any technical documentation you have which would provide a good starting place for understanding the code base / flow? Also, is there a specific ide i should use and or extensions that should be utilized? What machine setup is needed to run the source code?

Thanks!

**Reply by atdimitrov** [reply]
Hi @Bond-Addict.

> Is there any technical documentation you have which would provide a good starting place for understanding the code base / flow?

There was never need for one, so unfrotunately no. I can give you a birdseye view of the project that would give you a good starting point. We can delve deeper as needed. The product has 2 main parts - the UI and the decompilation engine.

The UI is contained almost entirely inside the `CodemerxDecompile` project. That's also the startup project for the entire solution. It is written using [Avalonia UI](https://avaloniaui.net) and generally follows the [MVVM pattern](https://docs.avaloniaui.net/docs/concepts/the-mvvm-pattern/avalonia-ui-and-mvvm). There are few places where this is not the case due to time constraints, but those are meant to be transitioned to MVVM eventially. The only other UI-related project is `AvaloniaEdit`, which is forked from the [official AvaloniaEdit repo](https://github.com/AvaloniaUI/AvaloniaEdit) to allow us to extend it beyond what's possible in the official AvaloniaEdit library.

All other projects are part of the decompilation engine and other auxiliary decompilation functions like project generation. Writing a good documentation for each one would be a very time consuming task and I don't think it is necessary at this point. We can go deeper as necessary.

> Also, is there a specific ide i should use and or extensions that should be utilized? What machine setup is needed to run the source code?

The solution is tested and works fine in both Visual Studio and Rider. As for extensions, there is one for both that enables designer support for Avalonia - [Avalonia for Visual Studio 2022](https://marketplace.visualstudio.com/items?itemName=AvaloniaTeam.AvaloniaVS) and [AvaloniaRider](https://plugins.jetbrains.com/plugin/14839-avaloniarider). As for machine setup - no special setup is required. The project currently targets .NET 7 so you will need the [.NET 7 SDK](https://dotnet.microsoft.com/en-us/download/dotnet/7.0), but apart from that no other setup should be necessary.

Happy to help with any problems you may encounter.

**Reply by atdimitrov** [reply]
One more thing to keep in mind is that the repo is using git submodules for both JustDecompileEngine and AvaloniaEdit. That makes cloning the repo a bit more involved. After you clone the repo you want to `git submodule init` to initialize both submodules and then `git submodule update` to checkout the correct commits.
